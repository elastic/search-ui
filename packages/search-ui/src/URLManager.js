import { createBrowserHistory as createHistory } from "history";
import queryString from "./queryString";

function isNumericString(num) {
  return !isNaN(num);
}

function toSingleValue(val) {
  return Array.isArray(val) ? val[val.length - 1] : val;
}

function toSingleValueInteger(num) {
  return toInteger(toSingleValue(num));
}

function toInteger(num) {
  if (!isNumericString(num)) return;
  return parseInt(num, 10);
}

function parseFiltersFromQueryParams(queryParams) {
  return queryParams.filters;
}

function parseCurrentFromQueryParams(queryParams) {
  return toSingleValueInteger(queryParams.current);
}

function parseSearchTermFromQueryParams(queryParams) {
  return toSingleValue(queryParams.q);
}

function parseSortFromQueryParams(queryParams) {
  const sortField = toSingleValue(queryParams["sort-field"]);
  const sortDirection = toSingleValue(queryParams["sort-direction"]);

  if (sortField) return [sortField, sortDirection];
  return [];
}

function parseSizeFromQueryParams(queryParams) {
  return toSingleValueInteger(queryParams.size);
}

function paramsToState(queryParams) {
  const state = {
    current: parseCurrentFromQueryParams(queryParams),
    filters: parseFiltersFromQueryParams(queryParams),
    searchTerm: parseSearchTermFromQueryParams(queryParams),
    resultsPerPage: parseSizeFromQueryParams(queryParams),
    sortField: parseSortFromQueryParams(queryParams)[0],
    sortDirection: parseSortFromQueryParams(queryParams)[1]
  };

  return Object.keys(state).reduce((acc, key) => {
    const value = state[key];
    if (value) acc[key] = value;
    return acc;
  }, {});
}

function stateToParams({
  searchTerm,
  current,
  filters,
  resultsPerPage,
  sortDirection,
  sortField
}) {
  const params = {};

  if (current > 1) params.current = current;
  if (searchTerm) params.q = searchTerm;
  if (resultsPerPage) params.size = resultsPerPage;
  if (filters && filters.length > 0) {
    params["filters"] = filters;
  }
  if (sortField) {
    params["sort-field"] = sortField;
    params["sort-direction"] = sortDirection;
  }

  return params;
}

function stateToQueryString(state) {
  return queryString.stringify(stateToParams(state));
}

/**
 * The URL Manager is responsible for synchronizing state between
 * SearchDriver and the URL. There are 3 main cases we handle when
 * synchronizing:
 *
 * 1. When the app loads, SearchDriver will need to
 * read the current state from the URL, in order to perform the search
 * expressed by the query string. `getStateFromURL` is used for this case.
 *
 * 2. When the URL changes as a result of `pushState` or `replaceState`,
 * SearchDriver will need to be notified and given the updated state, so that
 * it can re-run the current search. `onURLStateChange` is used for this case.
 *
 * 3. When state changes internally in the SearchDriver, as a result of an
 * Action, it will need to notify the URLManager of the change. `pushStateToURL`
 * is used for this case.
 */

export default class URLManager {
  constructor() {
    this.history = createHistory();
    this.lastPushSearchString = "";
  }

  /**
   * Parse the current URL into application state
   *
   * @return {Object} - The parsed state object
   */
  getStateFromURL() {
    return paramsToState(queryString.parse(this.history.location.search));
  }

  /**
   * Push the current state of the application to the URL
   *
   * @param {Object} state - The entire current state from the SearchDriver
   * @param {boolean} options
   * @param {boolean} options.replaceUrl - When pushing state to the URL, use history 'replace'
   * rather than 'push' to avoid adding a new history entry
   */
  pushStateToURL(state, { replaceUrl = false } = {}) {
    const searchString = stateToQueryString(state);
    this.lastPushSearchString = searchString;
    const navigationFunction = replaceUrl
      ? this.history.replace
      : this.history.push;
    navigationFunction({
      search: `?${searchString}`
    });
  }

  /**
   * Add an event handler to be executed whenever state is pushed to the URL
   *
   * @callback requestCallback
   * @param {Object} state - Updated application state parsed from the new URL
   *
   * @param {requestCallback} callback
   */
  onURLStateChange(callback) {
    this.unlisten = this.history.listen(location => {
      // If this URL is updated as a result of a pushState request, we don't
      // want to notify that the URL changed.
      if (`?${this.lastPushSearchString}` === location.search) return;

      // Once we've decided to return based on lastPushSearchString, reset
      // it so that we don't break back / forward button.
      this.lastPushSearchString = "";

      callback(paramsToState(queryString.parse(location.search)));
    });
  }

  tearDown() {
    this.unlisten();
  }
}
