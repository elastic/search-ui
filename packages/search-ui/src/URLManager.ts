import {
  createBrowserHistory as createHistory,
  createMemoryHistory,
  History
} from "history";
import queryString from "./queryString";
import { Filter, RequestState, SortOption } from "./types";

type QueryParams = {
  filters?: Filter[];
  current?: number;
  q?: string;
  size?: number;
  "sort-field"?: string;
  "sort-direction"?: string;
  sort?: SortOption[];
};

function isNumericString(num): boolean {
  return !isNaN(num);
}

function toSingleValue(val): string {
  return Array.isArray(val) ? val[val.length - 1] : val;
}

function toSingleValueInteger(num): number {
  return toInteger(toSingleValue(num));
}

function toInteger(num): number {
  if (!isNumericString(num)) return;
  return parseInt(num, 10);
}

function parseFiltersFromQueryParams(queryParams: QueryParams): Filter[] {
  return queryParams.filters;
}

function parseCurrentFromQueryParams(queryParams: QueryParams): number {
  return toSingleValueInteger(queryParams.current);
}

function parseSearchTermFromQueryParams(queryParams: QueryParams): string {
  return toSingleValue(queryParams.q);
}

function parseOldSortFromQueryParams(
  queryParams: QueryParams
): [string, string] | [] {
  const sortField = toSingleValue(queryParams["sort-field"]);
  const sortDirection = toSingleValue(queryParams["sort-direction"]);

  if (sortField) return [sortField, sortDirection];
  return [];
}

function parseSizeFromQueryParams(queryParams: QueryParams): number {
  return toSingleValueInteger(queryParams.size);
}

function parseSortFromQueryParams(queryParams: QueryParams): SortOption[] {
  return queryParams["sort"];
}

function paramsToState(queryParams: QueryParams): RequestState {
  const state = {
    current: parseCurrentFromQueryParams(queryParams),
    filters: parseFiltersFromQueryParams(queryParams),
    searchTerm: parseSearchTermFromQueryParams(queryParams),
    resultsPerPage: parseSizeFromQueryParams(queryParams),
    sortField: parseOldSortFromQueryParams(queryParams)[0],
    sortDirection: parseOldSortFromQueryParams(queryParams)[1],
    sortList: parseSortFromQueryParams(queryParams)
  };

  return Object.keys(state).reduce((acc, key) => {
    const value = state[key];
    if (value) acc[key] = value;
    return acc;
  }, {});
}

function stateToParams({
  searchTerm,
  current,
  filters,
  resultsPerPage,
  sortDirection,
  sortField,
  sortList
}: RequestState): QueryParams {
  const params: QueryParams = {};
  if (current > 1) params.current = current;
  if (searchTerm) params.q = searchTerm;
  if (resultsPerPage) params.size = resultsPerPage;
  if (filters && filters.length > 0) {
    params["filters"] = filters;
  }
  if (sortList && sortList.length > 0) {
    params["sort"] = sortList;
  } else if (sortField) {
    params["sort-field"] = sortField;
    params["sort-direction"] = sortDirection;
  }
  return params;
}

function stateToQueryString(state: RequestState): string {
  return queryString.stringify(stateToParams(state));
}

/**
 * The URL Manager is responsible for synchronizing state between
 * SearchDriver and the URL. There are 3 main cases we handle when
 * synchronizing:
 *
 * 1. When the app loads, SearchDriver will need to
 * read the current state from the URL, in order to perform the search
 * expressed by the query string. `getStateFromURL` is used for this case.
 *
 * 2. When the URL changes as a result of `pushState` or `replaceState`,
 * SearchDriver will need to be notified and given the updated state, so that
 * it can re-run the current search. `onURLStateChange` is used for this case.
 *
 * 3. When state changes internally in the SearchDriver, as a result of an
 * Action, it will need to notify the URLManager of the change. `pushStateToURL`
 * is used for this case.
 */

interface RoutingHandler {
  readUrl: () => string;
  writeUrl: (url: string, { replaceUrl }: { replaceUrl: boolean }) => void;
  urlToState: (url: string) => RequestState;
  stateToUrl: (state: RequestState) => string;
  routeChangeHandler: (handler: (url?: string) => void) => () => void;
}

export type RoutingHandlerOptions = Partial<RoutingHandler>;
type RoutingChangeCallback = (state: RequestState) => void;

export default class URLManager {
  history: History;
  lastPushSearchString: string;
  unlisten?: () => void;
  overrides: any;
  routingOptions: RoutingHandler;

  constructor(routingOptions: RoutingHandlerOptions = {}) {
    this.routingOptions = {
      readUrl: routingOptions.readUrl || this.readUrl.bind(this),
      writeUrl: routingOptions.writeUrl || this.writeUrl.bind(this),
      urlToState: routingOptions.urlToState || this.urlToState.bind(this),
      stateToUrl: routingOptions.stateToUrl || this.stateToUrl.bind(this),
      routeChangeHandler:
        routingOptions.routeChangeHandler || this.routeChangeHandler.bind(this)
    };

    this.history =
      typeof window !== "undefined" ? createHistory() : createMemoryHistory();
    this.lastPushSearchString = "";
  }

  /*
   * These functions are used to read and write the URL
   * Its designed to be overriden by the developer for their own 3rd party routing needs.
   * For example developers override this function to use next.js
   *
   **/
  readUrl() {
    return this.history ? this.history.location.search : "";
  }

  writeUrl(url: string, { replaceUrl = false }: { replaceUrl?: boolean } = {}) {
    const navigationFunction = replaceUrl
      ? this.history.replace
      : this.history.push;
    navigationFunction(`?${url}`);
  }

  /*
   * This function is used to convert a URL into a state object and vice versa
   * the state is stored as a search string in the URL.
   * Developers own implementations of this function should be able to handle full urls
   * and not just the search string.
   **/
  urlToState(url: string): RequestState {
    return paramsToState(queryString.parse(url));
  }

  stateToUrl(state: RequestState): string {
    return `${stateToQueryString(state)}`;
  }

  /**
   * Parse the current URL into application state
   *
   * @return {Object} - The parsed state object
   */
  getStateFromURL(): RequestState {
    return this.routingOptions.urlToState(this.routingOptions.readUrl());
  }

  /**
   * Push the current state of the application to the URL
   *
   * @param {Object} state - The entire current state from the SearchDriver
   * @param {boolean} options
   * @param {boolean} options.replaceUrl - When pushing state to the URL, use history 'replace'
   * rather than 'push' to avoid adding a new history entry
   */
  pushStateToURL(
    state: RequestState,
    { replaceUrl = false }: { replaceUrl?: boolean } = {}
  ): void {
    const url = this.routingOptions.stateToUrl(state);
    this.lastPushSearchString = url;
    this.routingOptions.writeUrl(url, { replaceUrl });
  }

  /**
   * Add an event handler to be executed whenever state is pushed to the URL
   *
   * @callback requestCallback
   * @param {Object} state - Updated application state parsed from the new URL
   *
   * @param {requestCallback} callback
   */
  onURLStateChange(callback: RoutingChangeCallback): void {
    const handler = (url) => {
      if (`?${this.lastPushSearchString}` === url) return;

      // Once we've decided to return based on lastPushSearchString, reset
      // it so that we don't break back / forward button.
      this.lastPushSearchString = "";

      callback(this.routingOptions.urlToState(url));
    };

    this.unlisten = this.routingOptions.routeChangeHandler(handler.bind(this));
  }

  routeChangeHandler(callback) {
    const handler = (location) => {
      callback(location.search);
    };
    return this.history.listen(handler);
  }

  tearDown(): void {
    this.unlisten();
  }
}
