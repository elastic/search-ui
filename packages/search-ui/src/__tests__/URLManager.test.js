import URLManager from "../URLManager";

function createManager() {
  const manager = new URLManager();
  manager.history = {
    location: {
      search: ""
    },
    listen: jest.fn(),
    push: jest.fn(),
    replace: jest.fn()
  };
  return manager;
}

const basicParameterState = {
  filters: [
    {
      dependencies: ["underscore", "another"]
    },
    {
      keywords: ["node"]
    }
  ],
  resultsPerPage: 20,
  searchTerm: "node",
  sortDirection: "asc",
  sortField: "name"
};

const basicParameterStateAsUrl =
  "?filters[0][dependencies][0]=underscore&filters[0][dependencies][1]=another&filters[1][keywords][0]=node&q=node&size=20&sort-direction=asc&sort-field=name";

const parameterStateWithRangeFilters = {
  filters: [
    {
      date: [
        {
          from: 12,
          to: 4000
        },
        {
          to: 4000
        }
      ]
    },
    {
      cost: [
        {
          from: 50
        }
      ]
    },
    {
      keywords: "node"
    }
  ]
};

it("can be initialized", () => {
  const manager = new URLManager();
  expect(manager).toBeInstanceOf(URLManager);
});

describe("#getStateFromURL", () => {
  it("will parse the current state from the URL", () => {
    const manager = createManager();
    manager.history.location.search =
      "?filters%5B0%5D%5Bdependencies%5D%5B0%5D=underscore&filters%5B1%5D%5Bkeywords%5D%5B0%5D=node&q=node&size=20&sort-direction=asc&sort-field=name";

    const state = manager.getStateFromURL();
    expect(state).toMatchSnapshot();
  });

  it("will parse range filter types", () => {
    const manager = createManager();
    manager.history.location.search =
      "?filters[0][date][0][from]=n_12_n&filters[0][date][0][to]=n_4000_n&filters[0][date][1][to]=n_4000_n&filters[1][cost][0][from]=n_50_n&filters[2][keywords][0]=node";

    const state = manager.getStateFromURL();
    expect(state).toMatchSnapshot();
  });

  it("will ignore unrecognized parameters", () => {
    const manager = createManager();
    manager.history.location.search =
      "?q=test&bogus=12&tommy=dunn&whatever&ok&thatsall";

    const state = manager.getStateFromURL();
    expect(state).toMatchSnapshot();
  });

  it("will correctly handle multiple instances of the same parameter", () => {
    const manager = createManager();
    manager.history.location.search =
      "filters[0][dependencies][0]=underscore&filters[0][dependencies][1]=another&filters[1][keywords][0]=node&q=bad&q=node&size=bad&size=20&sort-direction=bad&sort-direction=asc&sort-field=bad&sort-field=name";

    const state = manager.getStateFromURL();
    expect(state).toMatchSnapshot();
  });
});

describe("#pushStateToURL", () => {
  it("will update the url with the url corresponding to the provided state", () => {
    const manager = createManager();
    manager.pushStateToURL(basicParameterState);
    const queryString = manager.history.push.mock.calls[0][0].search;
    expect(queryString).toEqual(
      "?q=node&size=n_20_n&filters%5B0%5D%5Bdependencies%5D%5B0%5D=underscore&filters%5B0%5D%5Bdependencies%5D%5B1%5D=another&filters%5B1%5D%5Bkeywords%5D%5B0%5D=node&sort-field=name&sort-direction=asc"
    );
  });

  describe("filters", () => {
    it("will update the url with range filter types", () => {
      const manager = createManager();
      manager.pushStateToURL(parameterStateWithRangeFilters);
      const queryString = manager.history.push.mock.calls[0][0].search;
      expect(queryString).toEqual(
        "?filters%5B0%5D%5Bdate%5D%5B0%5D%5Bfrom%5D=n_12_n&filters%5B0%5D%5Bdate%5D%5B0%5D%5Bto%5D=n_4000_n&filters%5B0%5D%5Bdate%5D%5B1%5D%5Bto%5D=n_4000_n&filters%5B1%5D%5Bcost%5D%5B0%5D%5Bfrom%5D=n_50_n&filters%5B2%5D%5Bkeywords%5D=node"
      );
    });
  });

  describe("replaceUrl", () => {
    it("will update the url using 'replace' instead of 'push", () => {
      const manager = createManager();
      manager.pushStateToURL(basicParameterState, { replaceUrl: true });
      const queryString = manager.history.replace.mock.calls[0][0].search;
      expect(queryString).toEqual(
        "?q=node&size=n_20_n&filters%5B0%5D%5Bdependencies%5D%5B0%5D=underscore&filters%5B0%5D%5Bdependencies%5D%5B1%5D=another&filters%5B1%5D%5Bkeywords%5D%5B0%5D=node&sort-field=name&sort-direction=asc"
      );
    });
  });
});

describe("#onURLStateChange", () => {
  let manager;

  function setup() {
    manager = createManager();
  }

  function pushStateToURL(state) {
    manager.pushStateToURL(state);
    return manager.history.push.mock.calls[0][0].search;
  }

  function simulateBrowserHistoryEvent(newUrl) {
    // Since we're not in a real browser, we simulate the change event that
    // would have ocurred.
    manager.history.listen.mock.calls[0][0]({
      search: newUrl
    });
  }

  function subject(onURLStateChangeHandler, newUrl) {
    // Set our handler, which we be called any time url state changes
    manager.onURLStateChange(onURLStateChangeHandler);

    simulateBrowserHistoryEvent(newUrl);
  }

  it("will call provided callback with updated state when url changes", () => {
    setup();
    let newState;

    // Provide url state change handler
    subject(state => {
      newState = state;
    }, basicParameterStateAsUrl);

    // Verify it is called when url state changes
    expect(newState).toEqual(basicParameterState);
  });

  /*
    If we triggered the callback every time we pushed state to the url,
    we would have an infinite loop.

    pushStateToURL -> UpdateUrl -> HistoryEvent -> Handler -> pushstateToURL

    Instead, we just want:
    pushStateToURL -> UpdateUrl -> HistoryEvent -> Handler

  */
  it("will not trigger callback as a result of pushStateToURL", () => {
    setup();

    // Push state to the url
    // newUrl is the url that the browser changed to
    const newUrl = pushStateToURL(basicParameterState);
    let newState;

    // Provide url state change handler
    subject(state => {
      newState = state;
    }, newUrl);

    // Verify it is called when url state changes
    expect(newState).toBe(undefined);
  });

  it("will trigger callback if navigating back to a url with browser buttons", () => {
    setup();

    // Push state to the url
    // newUrl is the url that the browser changed to
    const newUrl = pushStateToURL(basicParameterState);
    let newState;

    // Provide url state change handler
    subject(state => {
      newState = state;
    }, newUrl);

    // Verify it is called when url state changes
    expect(newState).toBe(undefined);

    // Back button to empty state
    simulateBrowserHistoryEvent("");
    expect(newState).toEqual({}); // Was called, but empty state

    // Forward button to double back to the original "newUrl"
    simulateBrowserHistoryEvent(newUrl);
    expect(newState).toEqual(basicParameterState);
  });
});
