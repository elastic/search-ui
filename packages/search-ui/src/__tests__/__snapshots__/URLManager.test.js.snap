// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#getStateFromURL will correctly handle multiple instances of the same parameter 1`] = `
Object {
  "filters": Array [
    Object {
      "dependencies": Array [
        "underscore",
        "another",
      ],
    },
    Object {
      "keywords": Array [
        "node",
      ],
    },
  ],
  "resultsPerPage": 20,
  "searchTerm": "node",
  "sortDirection": "asc",
  "sortField": "name",
}
`;

exports[`#getStateFromURL will ignore unrecognized parameters 1`] = `
Object {
  "searchTerm": "test",
}
`;

exports[`#getStateFromURL will parse range filter types 1`] = `
Object {
  "filters": Array [
    Object {
      "date": Array [
        Object {
          "from": 12,
          "to": 4000,
        },
        Object {
          "to": 4000,
        },
      ],
    },
    Object {
      "cost": Array [
        Object {
          "from": 50,
        },
      ],
    },
    Object {
      "keywords": Array [
        "node",
      ],
    },
  ],
}
`;

exports[`#getStateFromURL will parse sortList items 1`] = `
Object {
  "sortList": Array [
    Object {
      "states": "asc",
    },
    Object {
      "title": "asc",
    },
  ],
}
`;

exports[`#getStateFromURL will parse the current state from the URL 1`] = `
Object {
  "filters": Array [
    Object {
      "dependencies": Array [
        "underscore",
      ],
    },
    Object {
      "keywords": Array [
        "node",
      ],
    },
  ],
  "resultsPerPage": 20,
  "searchTerm": "node",
  "sortDirection": "asc",
  "sortField": "name",
}
`;
